File: bots/python/expansion_bot.py
import os
import sys
from typing import Any

SDK_PATH = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "sdks", "python"))
if SDK_PATH not in sys.path:
    sys.path.insert(0, SDK_PATH)

from openforest_sdk import run_stdio


def bot(observation: dict[str, Any]) -> list[dict[str, Any]]:
    player_id = observation.get("player_id")
    planets = observation.get("planets", [])
    owned = [p for p in planets if p.get("owner") == player_id]
    neutrals = [p for p in planets if p.get("owner") is None]
    actions: list[dict[str, Any]] = []

    if not owned:
        return actions

    for source in sorted(owned, key=lambda p: p.get("energy", 0.0), reverse=True):
        if len(actions) >= observation.get("max_actions", 5):
            break
        if source.get("energy", 0.0) < source.get("energy_cap", 0.0) * 0.5:
            continue
        if not neutrals:
            break
        target = min(
            neutrals,
            key=lambda p: (p["x"] - source["x"]) ** 2 + (p["y"] - source["y"]) ** 2,
        )
        actions.append({
            "type": "send_fleet",
            "from_id": source["id"],
            "to_id": target["id"],
            "energy": max(8.0, source["energy"] * 0.35),
        })
    if len(actions) < observation.get("max_actions", 5):
        home = max(owned, key=lambda p: p.get("energy_cap", 0.0))
        actions.append({"type": "upgrade", "planet_id": home["id"], "upgrade": "energy"})
    return actions


if __name__ == "__main__":
    run_stdio(bot)

File: bots/node/expansion_bot.ts
import { runStdio } from "../../sdks/node/src/index";

runStdio((observation) => {
  const playerId = observation.player_id as number | null;
  const planets = (observation.planets ?? []) as Array<Record<string, any>>;
  const owned = planets.filter((p) => p.owner === playerId);
  const neutrals = planets.filter((p) => p.owner === null);
  const actions: Array<Record<string, unknown>> = [];

  if (!owned.length) {
    return actions;
  }

  const maxActions = Number(observation.max_actions ?? 5);
  const sortedOwned = [...owned].sort((a, b) => b.energy - a.energy);

  for (const source of sortedOwned) {
    if (actions.length >= maxActions || !neutrals.length) {
      break;
    }
    if (source.energy < source.energy_cap * 0.5) {
      continue;
    }
    const target = neutrals.reduce((best, p) => {
      const dist = (p.x - source.x) ** 2 + (p.y - source.y) ** 2;
      const bestDist = (best.x - source.x) ** 2 + (best.y - source.y) ** 2;
      return dist < bestDist ? p : best;
    }, neutrals[0]);
    actions.push({
      type: "send_fleet",
      from_id: source.id,
      to_id: target.id,
      energy: Math.max(8, source.energy * 0.35),
    });
  }

  if (actions.length < maxActions) {
    const home = owned.reduce((best, p) => (p.energy_cap > best.energy_cap ? p : best), owned[0]);
    actions.push({ type: "upgrade", planet_id: home.id, upgrade: "energy" });
  }

  return actions;
});
